### ISSUE TITLE
Implement Cloud Messaging Feature

### ISSUE THREAD

**Issue by simonhaenisch** [firebase/firebase-admin-dotnet#8]
We are in the process of adding Firebase into an existing project in order to add a real-time chat (using Firestore), and we would like to switch to FCM for push notifications. This would simplify sending chat-related notifications directly through Firebase, but we still need to send other notifications from the .NET backend, and thus need the Cloud Messaging feature.

I couldn't find a roadmap for the .NET Admin SDK so I'm just putting this out here as our most urgent feature request.

**Comment by hiranya911** [comment]
Thanks @simonhaenisch. Would you be willing to provide us a pull request implementing this feature? There are several reference implementations in other languages that we can base this on. I'd be happy to help with design and code reviews.

**Comment by simonhaenisch** [comment]
I can give it a try but only for dotnet core because I'm on macOS and don't have a dotnet development environment set up yet/anymore... I also have only little experience with dotnet, but with code review I guess I could give it a try. But if anyone else wants to take this on, please do so 🤓

**Comment by hiranya911** [comment]
We also do most of the development using .NET Core on Mac and Linux (we are in the process of setting up some CI for Windows and .NET Framework).

**Comment by simonhaenisch** [comment]
@hiranya911 could you please elaborate how I can retrieve the ID token from the Firebase app instance? In the Node.js package `this.app.INTERNAL.getToken()` is used to set the `Authorization: Bearer ...` header for requests to the API.

I also tried looking into the Node.js and Python implementations for Firebase and saw that they build on top of Google Cloud Platform packages. So would the recommended way of implementing Firestore in this Admin SDK be to build on top of <https://googlecloudplatform.github.io/google-cloud-dotnet/>?

**Comment by hiranya911** [comment]
ID tokens are a client-side concept, and cannot be easily retrieved in the server-side. Assuming you meant access tokens, you can use [`GoogleCredential`](https://developers.google.com/api-client-library/dotnet/reference/1.9.3/classGoogle_1_1Apis_1_1Auth_1_1OAuth2_1_1GoogleCredential) class to get hold of them. We already have code that makes OAuth2 authorized HTTP requests in this manner:

https://github.com/firebase/firebase-admin-dotnet/blob/c57713f4c86da38bfdee4a2ea665be278f56072e/FirebaseAdmin/FirebaseAdmin/Auth/IAMSigner.cs#L47

Not sure how Firestore is relevant to this thread, but there's https://www.nuget.org/packages/Google.Cloud.Firestore/1.0.0-beta09 for that.

**Comment by DenSmoke** [comment]
@simonhaenisch I am sending messages via FCM this way:
```C#
using Google.Apis.Auth.OAuth2;
using Google.Cloud.Firestore;
using Newtonsoft.Json;
///
....
///
private ServiceAccountCredential _saCredential;
private async Task<bool> AuthorizeAsync()
{
   var appDefault = await GoogleCredential.GetApplicationDefaultAsync().ConfigureAwait(false);
   _saCredential = (ServiceAccountCredential)appDefault.CreateScoped(_scopes).UnderlyingCredential;
   return await _saCredential.RequestAccessTokenAsync(CancellationToken.None).ConfigureAwait(false);
}
public async Task PushMessageAsync(object message)
{
            if (_saCredential.Token.IsExpired(_saCredential.Clock))
                await _saCredential.RequestAccessTokenAsync(CancellationToken.None).ConfigureAwait(false);
            string json = JsonConvert.SerializeObject(message, Formatting.None,
            new JsonSerializerSettings
            {
                NullValueHandling = NullValueHandling.Ignore
            });
            _reciever?.Invoke(json);
            var webRequest = WebRequest.Create(_fcmEndpoint);
            webRequest.Method = "POST";
            webRequest.ContentType = "application/json; charset=UTF-8";
            webRequest.Headers.Add($"Authorization: {_saCredential.Token.TokenType} {_saCredential.Token.AccessToken}");
            byte[] buffer = Encoding.UTF8.GetBytes(json);
            webRequest.ContentLength = buffer.Length;
            using (var dataStream = await webRequest.GetRequestStreamAsync().ConfigureAwait(false))
                await dataStream.WriteAsync(buffer, 0, buffer.Length).ConfigureAwait(false);
            using (var response = await webRequest.GetResponseAsync().ConfigureAwait(false))
            using (var responseStream = response.GetResponseStream())
                try
                {
                    using (var dataStreamReader = new StreamReader(responseStream))
                        _reciever?.Invoke((await dataStreamReader.ReadToEndAsync().ConfigureAwait(false)).Replace("\n", "").Replace("\r", ""));
                }
                catch (Exception ex)
                {
                    _reciever?.Invoke(ex.Message);
                }
}
```
Also I have set environment variable
```
set GOOGLE_APPLICATION_CREDENTIALS=path\to\key.json
````
May be this example will help you

**Comment by theberserker** [comment]
@simonhaenisch, @hiranya911 How far is this? I see lots of code implemented already. Is this tested and might be released soon?

**Comment by simonhaenisch** [comment]
@theberserker I haven't done anything regarding this anymore because we ended up not using FCM for now (at least not directly)... but yeah, seems like hiranya911 started working on this recently and implemented the basic stuff in #12 already :tada: I'll close this feature request now as I expect this to be released soonish.
